src/mcpwm/timer.rs

```diff
@@ -0,0 +1,302 @@
+    ///
+    /// The wave form will end up looking something like the following:
+    /// ```
+    ///                                             change count dir to decrement, counter = period
+    ///       start, counter = 0, incrementing          |                                     change count dir to increment, counter = 0
+    ///         |                                       |                                        |
+    ///         |                                       |*--- counter = period                   |*----- start, counter = 0, incrementing
+    ///         v <----  duty  ----> .                  v|                  . <----  duty  ----> ||
+    ///         .                    .                  .v                  .                    vv
+    ///         ---------------------.                  ..                  .-------------------------------------------.                  ..                  .--
+    ///                 Active       |                  ..                  |        Active                Active       |                  ..                  |
+    ///                              |                  ..                  |                                           |                  ..                  |
+    ///                              |     Not active   ..    Not active    |                                           |     Not active   ..    Not active    |
+    ///                              ----------------------------------------                                           ----------------------------------------
+    /// ```
+    /// NOTE: That in this mode, the frequency will be half of that specified
+    UpDown,
+
+    /// Timer paused
+    Pause,
+}
+
+impl From<CountMode> for mcpwm_timer_count_mode_t {
+    fn from(val: CountMode) -> Self {
+        match val {
+            //CounterMode::Frozen => mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+            CountMode::Up => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP,
+            CountMode::Down => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_DOWN,
+            CountMode::UpDown => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP_DOWN,
+            CountMode::Pause => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_PAUSE,
+        }
+    }
+}
+
+pub struct TIMER<const N: u8, G: Group> {
+    _ptr: PhantomData<*const ()>,
+    _group: PhantomData<G>,
+}
+
+impl<const N: u8, G: Group> TIMER<N, G> {
+    /// # Safety
+    ///
+    /// Care should be taken not to instnatiate this peripheralinstance, if it is already instantiated and used elsewhere
+    #[inline(always)]
+    pub unsafe fn new() -> Self {
+        Self {
+            _ptr: PhantomData,
+            _group: PhantomData,
+        }
+    }
+}
+
+unsafe impl<const N: u8, G: Group> Send for TIMER<N, G> {}
+
+impl<const N: u8, G: Group> crate::peripheral::sealed::Sealed for TIMER<N, G> {}
+
+impl<const N: u8, G: Group> crate::peripheral::Peripheral for TIMER<N, G> {
```

Comment:
Just to note that implementing `Peripheral` only makes sense if you allow - in the `TimerDriver::new` constructor - to pass the peripheral not just by moving, but also by `&mut ref`, which is not possible right now. I'll elaborate further down in the code how that can become possible.